[docs] usecase: opentelemetry (#3439)
feldera/feldera: docs/use_cases/otel/grafana.md

# Grafana Integration

> Next step is to integrate with Grafana so that we can query Feldera for these insights.

Reviewer: we don't query Feldera, supposedly Grafana does.
Explain what the flow is. 
I expected that Feldera would push changes to grafana, but perhaps you are querying a materialized view in Feldera?
(You could talk briefly about these two approaches)

We plan to create a Grafana dashboard like the one shown below:

![Grafana Dashboard](grafana-dashboard.png)

> This dashboard takes inspiration from the workshop: [Analyzing and Visualizing OpenTelemetry Traces with SQL - John Pruitt, Timescale](https://www.youtube.com/watch?v=rjJgZ_MNmPQ)

Reviewer: Make this a footnote or a bibliographic reference at the end.

#### Step 1: Install the Feldera Grafana Plugin

> Since the Feldera datasource plugin is not yet available in the Grafana marketplace, it must be **manually installed** from [GitHub](https://github.com/feldera/grafana-datasource). You can download the latest release and unzip in in the plugins directory of your Grafana server.

Reviewer: say what the plugin does

7. Open the Grafana Dashboard and go to **Connections -> Add New Connection**
8. Search for **Feldera** and select the **Feldera Data Source Plugin**.
9. Click **"Add a new data source"**.

#### Step 3: Configure the Feldera Data Source

![Feldera Data Source Config](feldera-data-source.png)

##### Required Fields:

| Field         | Value                                                            |
| ------------- | ---------------------------------------------------------------- |
| Base URL      | `http://feldera:8080` _(If running in Docker alongside Grafana)_ |
| Pipeline Name | `otel` _(Your Feldera pipeline name)_                            |
| API Key       | _(If authentication is required, provide the API Key)_           |

Click **"Save & Test"** to verify the connection.

## Creating a Grafana Dashboard

Once Feldera is set up as a data source, you can build a **custom Grafana dashboard** to visualize OpenTelemetry data.

For all the visualizations we create here, use the **feldera** data source.

#### Requests Count

This visualization displays the total number of requests in the selected time range.
Visualization Type: **Stat**

```sql

> -- requests count

Reviewer: who runs this query?
this is parameterized SQL.
it seems that Grafana runs this.

WHERE
	s.parentSpanId = '' AND
	s.eventTime BETWEEN $__timeFrom() AND $__timeTo()
ORDER BY
	elapsedTimeMillis DESC
LIMIT
	6
```
#### Histogram of Latencies

Provides a latency histogram to analyze response time patterns.
Visualization Type: **Histogram**

```sql
-- histogram of latencies
SELECT
	s.elapsedTimeMillis
FROM
	spans s
WHERE
	s.parentSpanId = '' AND
	s.eventTime BETWEEN $__timeFrom() AND $__timeTo()
ORDER BY
	s.eventTime DESC
```

#### P95 Latencies

Tracks the **95th percentile latency** over time.
Visualization Type: **"Bar chart"**

```sql
-- p95_latency
SELECT
	*
FROM
	p95_latency
WHERE
	time BETWEEN $__timeFrom() AND $__timeTo()
ORDER BY
	time DESC
```

#### Throughput

Measures the **number of requests per 10 second buckets**.
Visualization Type: **"Bar chart"**

```sql
-- throughput
SELECT
	*
FROM
	throughput
WHERE
	time BETWEEN $__timeFrom() AND $__timeTo()
ORDER BY
	time DESC
```

#### Pie chart of Operation Executions Time

Displays **execution time per operation**, showing where time is spent.
Visualization Type: **"Pie chart"**

```sql
-- operation executions time
SELECT
	sum(elapsed) as elapsed,
	name
FROM
	operation_execution_time
GROUP BY
	name
```

> 

Reviewer: you need a conclusion

"look how easy it was, etc."
any idea about performance or memory consumption?
this program may benefit from LATENESS annotations to GC old state.
although materialized views may require an additional WHERE clause.

# Grafana Integration

> Next step is to integrate with Grafana so that we can query Feldera for these insights.

Reviewer 2: ```suggestion
The next step is to integrate with Grafana so that we can query Feldera for these insights.
```

> We plan to create a Grafana dashboard[^1] like the one shown below:

Reviewer 2: ```suggestion
We will create a Grafana dashboard[^1] like the one shown below:
```

![Grafana Dashboard](grafana-dashboard.png)

[^1]: This dashboard takes inspiration from the workshop: [Analyzing and Visualizing OpenTelemetry Traces with SQL - John Pruitt, Timescale](https://www.youtube.com/watch?v=rjJgZ_MNmPQ)

#### Step 1: Install the Feldera Grafana Plugin

Since the Feldera data source plugin is not yet available in the Grafana marketplace, it must be **manually installed** from [GitHub](https://github.com/feldera/grafana-datasource). You can download the latest release and unzip in in the plugins directory of your Grafana server.

> This plugin allows Grafana to seamlessly execute ad-hoc queries on Feldera pipelines, to treat Feldera just like any other compatible data source.

Reviewer 2: reference to ad hoc queries docs
